### PR TITLE
Add address diff functionality

### PR DESCRIPTION
**Pull request by ChrisSchinnerl**
# PULL REQUEST

## Overview

Adds support for fetching the diff between the addresses that skyd has and the ones in the db.

## Checklist

- [ ] All git commits are signed. (REQUIRED)
- [ ] All new methods or updated methods have clear docstrings.
- [ ] Testing added or updated for new methods.
- [ ] Verified if any changes impact the WebPortal Health Checks.
- [ ] Appropriate documentation updated.
- [ ] Changelog file created.

### CODE REVIEW

**Review by ro-tex — APPROVED** [review]

File: promoter/database_test.go [hunk 1]
```diff
@@ -83,18 +28,15 @@ func TestAddressWatcher(t *testing.T) {
...
-		case "insert":
-			inserted[update.DocumentKey.Address] = struct{}{}
-		case "delete":
-			deleted[update.DocumentKey.Address] = struct{}{}
-		case "drop":
-		case "invalidate":
+		case operationTypeInsert:
+			inserted[update.Address] = struct{}{}
+		case operationTypeDelete:
+			deleted[update.Address] = struct{}{}
```
[the remark this answers is not in the document]
Comment (MSevey) [reply promoter/database_test.go:38]: good for future proofing tests

File: promoter/promoter.go [hunk 2]
```diff
@@ -85,3 +86,42 @@ func (p *Promoter) initBackgroundThreads(f updateFunc) {
 		p.threadedAddressWatcher(p.bgCtx, f)
 	}()
 }
+
+// staticAddrDiff returns a diff of addresses that describes which addresses
+// need to be added and removed from skyd to match the state of the database.
```
Comment (MSevey) [inline promoter/promoter.go:91]: Follow Up: might be worth adding some additional context here on the design. Initially I would have thought that skyd is the source of truth for address, and so it is the db that needs to be updated. 

Is there a situation where that diff is needed? And so we should have two separate diff functions depending on the diff needed?

**Review by MSevey — APPROVED** [review]

File: promoter/promoter.go (hunk 2 above)
Comment (ChrisSchinnerl) [reply promoter/promoter.go:91]: Added another sentence to make it clear that we expect skyd to match the state of the database.

We never need to update the database according to the state in skyd. Only the other way round.

File: promoter/database_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ChrisSchinnerl) [reply promoter/database_test.go]: Added. I could've sworn I saw 2 other types "drop" and "invalidate" but it doesn't seem to panic so I guess it's fine.

File: promoter/database.go [hunk 3]
```diff
@@ -108,19 +154,36 @@ OUTER:
...
+			continue OUTER              // try again
+		}
+		for _, addr := range toAdd {
+			f(WatchedAddressUpdate{
+				Address:       addr,
+				OperationType: operationTypeInsert,
+			})
+		}
+		for _, addr := range toRemove {
+			f(WatchedAddressUpdate{
```
[the remark this answers is not in the document]
Comment (ChrisSchinnerl) [reply promoter/database.go]: done

File: promoter/database.go [hunk 4]
```diff
@@ -108,19 +154,36 @@ OUTER:
...
+		// Fetch the diff of watched addresses and send updates down the
+		// callback accordingly.
+		toAdd, toRemove, err := p.staticAddrDiff(ctx)
+		if err != nil {
+			p.staticLogger.WithError(err).Error("Failed to fetch address diff")
+			time.Sleep(2 * time.Second) // sleep before retrying
+			continue OUTER              // try again
+		}
+		for _, addr := range toAdd {
+			f(WatchedAddressUpdate{
```
[the remark this answers is not in the document]
Comment (ChrisSchinnerl) [reply promoter/database.go]: done

File: promoter/database.go [hunk 5]
```diff
@@ -87,6 +115,24 @@ func (p *Promoter) staticColWatchedAddresses() *mongo.Collection {
 	return p.staticDB.Collection(colWatchedAddressesName)
 }
 
+// staticWatchedDBAddresses returns all watched addresses as seen in the
+// database.
+func (p *Promoter) staticWatchedDBAddresses(ctx context.Context) ([]types.UnlockHash, error) {
+	c, err := p.staticColWatchedAddresses().Find(ctx, bson.D{})
```
[the remark this answers is not in the document]
Comment (ChrisSchinnerl) [reply promoter/database.go:121]: Yeah at 1,000,000 subscribers with 1 address each we would fetch around 32MiB of addresses. So we probably don't need paging for now.

**Review by ro-tex — APPROVED** [review]

**Pull request merged**